File: GasProject Server/Sources/World/Objects/Projectile.cpp
#include "Projectile.hpp"

#include "World/Tile.hpp"
#include "World/Objects.hpp"

Projectile::Projectile(uf::vec2i direction) :
	startTile(nullptr)
{
    name = "Stun Orb";
    sprite = "stunorb";
    density = false;
    layer = 100;
    SetConstSpeed(direction.normalize() * speed);
}

void Projectile::Update(std::chrono::microseconds timeElapsed) {
    Object::Update(timeElapsed);

	if (CheckIfJustCreated())
		startTile = GetTile();

	if (!GetTile())
		throw std::exception(); // "Unexpected: projectile's tile is null!"

    if (startTile != GetTile() && GetTile()->IsDense()) {
        onHit(GetTile()->GetDenseObject());
        Delete();
    }
}

void Projectile::onHit(Object *obj) {
    if (auto *creature = dynamic_cast<Creature *>(obj))
        creature->Stun();
}

File: GameLogic/Objects/Projectile.py
from Engine.Geometry import DirectionSet, Direction
from Engine.World import Object

from Objects.Creature import Creature

class Projectile(Object):
	defName = "Stun Orb"
	defSprite = "stunorb"
	defDesc = "Taser's projectile. Watch out!"

	def __init__(self):
		super().__init__()
		self.layer = 100
		self.density = True
		self.startTile = None
		self.__absoluteSpeed = 7.0

	def BumpedTo(self, object):
		super().BumpedTo(object)
		self.__hit(object)
		return True

	def SetShotDirection(self, direction):
		direction = direction.Normalize()
		self.speed = direction * self.__absoluteSpeed

	def __hit(self, hittedObject):
		if isinstance(hittedObject, Creature):
			hittedObject.Stun()
		self.Delete()
